Composer/packages/lib/indexers/src/dialogUtils/jsonDiff.ts

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+import has from 'lodash/has';
+import get from 'lodash/get';
+import set from 'lodash/set';
+import isEqual from 'lodash/isEqual';
+import cloneDeep from 'lodash/cloneDeep';
+
+import { JsonWalk, VisitorFunc } from '../utils/jsonWalk';
+
+import { ListDiff } from './listDiff';
+
+export interface IJSONChangeAdd {
+  path: string;
+  value: any;
+}
+export interface IJSONChangeDelete {
+  path: string;
+  value: any;
+}
+export interface IJSONChangeUpdate {
+  path: string;
+  value: any;
+  preValue: any;
+}
+
+export interface IJsonChanges {
```

Comment:
This is not a good interface, usually you don't see change`s` as interface name, because you can use built-in types for this. 

I think if you want to make it generic, 

you would expect a Differ will diff two json and returns a set of changes. 

interface JsonDiffer {
 differ(jsonA, jsonB) :=> JsonDiff[]
}

in this definiton, there is only minimal concept that, a json differ diff json into diffs. that's it, as high level.

at low level, is you may have different type of Diff, that's totally a low level differend kind of type of JsonDiff.

all those above is just showing what mean by high-level vs low-level abstraction. But in our case, we probabaly don't need any of them.